lib/TPP/Transforms/HoistVectorTransfers.cpp

```diff
@@ -0,0 +1,198 @@
+
+struct HoistVectorTransferOp : OpRewritePattern<vector::ContractionOp> {
+  using OpRewritePattern<vector::ContractionOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::ContractionOp contractOp,
+                                PatternRewriter &rewriter) const override {
+
+        // Check whether the linalg tiling + vector contract pattern matches for the 4-nested loop structure
+        auto oldKForOp =  contractOp->getParentOfType<scf::ForOp>();
+        if (!oldKForOp)
+                return rewriter.notifyMatchFailure(contractOp, "Not a linalg tile + vector contract operation");
+
+        auto oldReductionForOp = oldKForOp->getParentOfType<scf::ForOp>();
+        if (!oldReductionForOp)
+                return rewriter.notifyMatchFailure(contractOp, "Not a linalg tile + vector contract operation");
+
+        auto oldNForOp = oldReductionForOp->getParentOfType<scf::ForOp>();
+        if (!oldNForOp)
+                return rewriter.notifyMatchFailure(contractOp, "Not a linalg tile + vector contract operation");
+
+        auto oldMForOp = oldNForOp->getParentOfType<scf::ForOp>();
+        if (!oldMForOp)
+                return rewriter.notifyMatchFailure(contractOp, "Not a linalg tile + vector contract operation");
+
+
+        // Check the vector contract operation satisfies the required pattern.
+        // Check the ACC, LHS, and RHS of contract operation
+        auto vectorReadOpAcc = contractOp.getAcc().getDefiningOp<mlir::vector::TransferReadOp>();
+        if (!vectorReadOpAcc)
+                return failure();
+
+        auto subviewOp = vectorReadOpAcc.getOperand(0).getDefiningOp<memref::SubViewOp>();
+        if (!subviewOp)
+                return failure();
+
+        auto vectorReadOpLHS = contractOp.getLhs().getDefiningOp<mlir::vector::TransferReadOp>();
+        if (!vectorReadOpLHS)
+                return failure();
+
+        auto vectorReadOpRHS = contractOp.getRhs().getDefiningOp<mlir::vector::TransferReadOp>();
+        if (!vectorReadOpRHS)
+                return failure();
+
+
+
+        // Check the operation type MatMul, B-MatMul, or BR-MatMul
+        auto contractIteratorTypes = contractOp.getIteratorTypesArray();
+        MatMulType matmulType;
+
+        if (contractIteratorTypes.size() > 3) {
+                matmulType = contractIteratorTypes[contractIteratorTypes.size() - 4] == vector::IteratorType::parallel ? MatMulType::Batch : MatMulType::BatchReduce;
+                if (matmulType == MatMulType::Batch)
+                        return rewriter.notifyMatchFailure(contractOp, "Batch matmul not supported");
+        } else if (contractIteratorTypes.size() == 3) {
+                matmulType = MatMulType::Standard;
+        } else {
+                return rewriter.notifyMatchFailure(contractOp, "The vector contract operation is not a gemm");
+        }
+
+        auto vectorReadOpLHSType = cast<ShapedType>(vectorReadOpLHS.getType());
```

Comment:
this is not saving a lot of typing. Since you're mostly using `getRank`, you can cache that instead.

Reply:
For `vectorReadOpLHSType`  we are later calling `getDimSize()` to do a check on the `k-dim`.
For `vectorReadOpRHSType` we do only `getRank()`, So I cached it.